File: sdks/js/langchain/cases/2-multi-step.js
/**
 * 2-multi-step: Multi-step Conversation
 *
 * Tests a multi-step conversation with conversation history using LangChain SDK
 * and verifies that Sentry captures all spans for both API calls.
 */

const { Sentry } = require("../setup");
const { ChatOpenAI } = require("@langchain/openai");
const { HumanMessage, SystemMessage, AIMessage } = require("@langchain/core/messages");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

async function testLogic(inputs) {
  const { model, system, first_prompt, second_prompt } = inputs;

  const chatModel = new ChatOpenAI({
    modelName: model,
    apiKey: process.env.OPENAI_API_KEY,
  });

  // First call
  const firstMessages = [
    new SystemMessage(system),
    new HumanMessage(first_prompt),
  ];

  const firstResponse = await chatModel.invoke(firstMessages);
  const firstText = firstResponse.content;

  if (!firstText) {
    throw new Error("No completion returned from LangChain (first call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    First response: ${firstText}`);
  }

  // Second call with conversation history
  const secondMessages = [
    new SystemMessage(system),
    new HumanMessage(first_prompt),
    new AIMessage(firstText),
    new HumanMessage(second_prompt),
  ];

  const secondResponse = await chatModel.invoke(secondMessages);
  const secondText = secondResponse.content;

  if (!secondText) {
    throw new Error("No completion returned from LangChain (second call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Second response: ${secondText}`);
  }
}

module.exports = runTestCase("2-multi-step", testLogic, Sentry);

File: sdks/py/langchain/cases/2-multi-step.py
"""
2-multi-step: Multi-step Conversation

Tests a multi-step conversation with conversation history using LangChain SDK
and verifies that Sentry captures all spans for both API calls.
"""

import os
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage, AIMessage
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]
    system = inputs["system"]
    first_prompt = inputs["first_prompt"]
    second_prompt = inputs["second_prompt"]

    chat = ChatOpenAI(model=model, api_key=os.getenv("OPENAI_API_KEY"))

    # First call
    first_messages = [
        SystemMessage(content=system),
        HumanMessage(content=first_prompt),
    ]

    first_response = chat.invoke(first_messages)
    first_text = first_response.content

    if not first_text:
        raise Exception("No output returned from LangChain (first call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    First response: {first_text}")

    # Second call with conversation history
    second_messages = [
        SystemMessage(content=system),
        HumanMessage(content=first_prompt),
        AIMessage(content=first_text),
        HumanMessage(content=second_prompt),
    ]

    second_response = chat.invoke(second_messages)
    second_text = second_response.content

    if not second_text:
        raise Exception("No output returned from LangChain (second call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Second response: {second_text}")

    return second_text


# Export test case functions
test_case = run_test_case("2-multi-step", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]
